lib/blocks.php

```diff
@@ -296,3 +296,197 @@ function gutenberg_enqueue_editor_block_styles_assets() {
+	$has_custom_background_color = array_key_exists( 'style', $attributes )
+	&& array_key_exists( 'color', $attributes['style'] )
+	&& array_key_exists( 'background', $attributes['style']['color'] );
+
+	// Gradient color.
+	$has_named_gradient  = array_key_exists( 'gradient', $attributes );
+	$has_custom_gradient = array_key_exists( 'style', $attributes )
+	&& array_key_exists( 'color', $attributes['style'] )
+	&& array_key_exists( 'gradient', $attributes['style']['color'] );
+
+	// If has background color.
+	if ( $has_custom_background_color || $has_named_background_color || $has_named_gradient || $has_custom_gradient ) {
+		// Add has-background class.
+		$background_colors['css_classes'][] = 'has-background';
+	}
+
+	if ( $has_named_background_color ) {
+		// Add the background-color class.
+		$background_colors['css_classes'][] = sprintf( 'has-%s-background-color', $attributes['backgroundColor'] );
+	} elseif ( $has_custom_background_color ) {
+		// Add the custom background-color inline style.
+		$background_colors['inline_styles'] .= sprintf( 'background-color: %s;', $attributes['style']['color']['background'] );
+	} elseif ( $has_named_gradient ) {
+		$background_colors['css_classes'][] = sprintf( 'has-%s-gradient-background', $attributes['gradient'] );
+	} elseif ( $has_custom_gradient ) {
+		$background_colors['inline_styles'] .= sprintf( 'background: %s;', $attributes['style']['color']['gradient'] );
+	}
+
+	return array(
+		'text'       => $text_colors,
+		'background' => $background_colors,
+	);
+}
+
+/**
+ * Build an array with CSS classes and inline styles defining the font sizes
+ * which will be applied to the block markup in the front-end.
+ *
+ * @param  array $attributes block attributes.
+ * @return array Font size CSS classes and inline styles.
+ */
+function gutenberg_experimental_build_css_typography( $attributes ) {
+	// CSS classes.
+	$typography = array(
+		'css_classes'   => array(),
+		'inline_styles' => '',
+	);
+
+	$has_named_font_size  = array_key_exists( 'fontSize', $attributes );
+	$has_custom_font_size = array_key_exists( 'style', $attributes )
+	&& array_key_exists( 'typography', $attributes['style'] )
+	&& array_key_exists( 'fontSize', $attributes['style']['typography'] );
+
+	if ( $has_named_font_size ) {
+		// Add the font size class.
+		$typography['css_classes'][] = sprintf( 'has-%s-font-size', $attributes['fontSize'] );
+	} elseif ( $has_custom_font_size ) {
+		// Add the custom font size inline style.
+		$typography['inline_styles'] .= sprintf( 'font-size: %spx;', $attributes['style']['typography']['fontSize'] );
+	}
+
+	$has_line_height = array_key_exists( 'style', $attributes )
+	&& array_key_exists( 'typography', $attributes['style'] )
+	&& array_key_exists( 'lineHeight', $attributes['style']['typography'] );
+
+	if ( $has_line_height ) {
+		$typography['inline_styles'] .= sprintf( 'line-height: %s;', $attributes['style']['typography']['lineHeight'] );
+	}
+
+	return $typography;
+}
```

Comment:
PHP unit tests in the repo are a bit tricky to run locally... but this definitely feels like a great candidate for unit tests.

Reply:
Where would the tests for this sit? In `phpunit/class-block-test.php`?

Reply:
I started a test, although It's still very much WIP (mostly trying to familiarize at this point).

> Where would the tests for this sit? In phpunit/class-block-test.php?

I could use some clarity on this, but maybe it makes sense to start a new file for these tests?

Currently, the test I started is in `class-block-context-test.php`, its probably not the appropriate place (?) but I put it there for now to take advantage of its `register_block_type` method.  

I did try in `class-wp-block-test.php` first, but the `registry->register` set up there doesn't seem sufficient for the `WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );` used in our function to know about the registered block in the test.  However the `register_block_type` method in the context test does seem sufficient.  

Maybe we want to set up another test class/file similar to that to run tests for this?  🤔